icepyx/core/visualization.py

```diff
@@ -0,0 +1,341 @@
+"""
+Interactive visualization of spatial extent and ICESat-2 elevations
+"""
+import intake.source.utils
+import backoff
+import requests
+import numpy as np
+import pandas as pd
+from itertools import compress
+import concurrent.futures
+from tqdm import tqdm
+import xarray as xr
+import dask.array as da
+import datashader as ds
+import geoviews as gv
+import holoviews as hv
+from holoviews.operation.datashader import rasterize
+
+import icepyx as ipx
+
+hv.extension('bokeh')
+gv.extension('bokeh')
+
+
+def files_in_latest_n_cycles(files, cycles, n=1):
+    """
+    Get list of file names in latest n cycles
+    """
+    if n == 1:
+        latest_cycle = max(cycles)
+        viz_file_list = [f for f in files if int(f.rsplit('_')[-3][4:6]) == latest_cycle]
+        return viz_file_list
+
+    elif n > 1:
+        if len(cycles) >= n:
+            viz_cycle_list = [max(cycles) - i for i in np.arange(n)]
+        else:
+            viz_cycle_list = cycles
+        viz_file_list = [f for f in files if int(f.rsplit('_')[-3][4:6]) in viz_cycle_list]
+        return viz_file_list
+
+    else:
+        raise Exception('Wrong n value')
+
+
+class Visualize:
+
+    def __init__(self, product, bbox, date_range=None, cycles=None, tracks=None):
+        self.product = product
+        self.bbox = bbox
+        self.date_range = date_range
+        self.cycles = cycles
+        self.tracks = tracks
```

Comment:
I remember we talked at an icepyx meeting about making it so that the user could call `Visualize()` or `Query.visualize()`. With that in mind, I'm thinking that we should use the `__init__` to allow both inputs and end up with a Visualize object that has the required info that's been validated (so if the user enters the search criteria directly, `query_icesat2_filelist` is called and we thus know that if a Visualize object has been created it has a valid file list to do the visualizations). I'm happy to work on this cross-object synchronization, @icetianli.

Reply:
Thanks @JessicaS11 for checking on this! Yes that's what we discussed in our last meeting, I agree the ultimate goal should be directly calling `Visualize` in `Query` class, it's a good idea to use `query_icesat2_filelist` to validate if the searched region has data or not so we can inform users of the results. The only problem is `Visualize` class needs to create an individual Query class object for each 5*5 bounding box to query OA API, so I am not very sure what's the best way to connect `Query` and `Visualize` classes, I'll check later.

Reply:
Ah - that makes sense. I'll give it more thought too (I think I was misunderstanding exactly what `query_icesat2_filelist` does), but at the moment I don't think creating multiple query objects within the code (since the user never sees them) will be an issue.

I'm thinking the mechanisms for access could be something like:
1. `reg_a_vis = ipx.Query(product, bbox, date/cycles/tracks).visualize()`
2. `reg_a_vis = ipx.Visualize([inputs])` where inputs is either a) an existing query object; b) product, bbox, date/cycles/tracks. In the former case, if the user inputs a shapefile or polygon, we could turn it into a bbox using `.bounds()`.

Reply:
Now `Query` should be able to call `Visualize` directly via `visualize_elevation` in `Query` class, see https://github.com/icesat2py/icepyx/pull/144/commits/0911ea45976490464af759136388d87402750827, I also converted the polygon/ shapefile input into bounding box by using `[shapely.minimum_rotated_rectangle](https://shapely.readthedocs.io/en/stable/manual.html)`, so non-bounding box input can also be used for elevation check now.